modal/file_io.py

```diff
@@ -0,0 +1,375 @@
+# Copyright Modal Labs 2024
+import asyncio
+import io
+from typing import TYPE_CHECKING, AsyncIterator, Generic, Optional, Sequence, TypeVar, Union
+
+if TYPE_CHECKING:
+    import _typeshed
+
+from grpclib.exceptions import GRPCError, StreamTerminatedError
+
+from modal._utils.grpc_utils import retry_transient_errors
+from modal_proto import api_pb2
+
+from ._utils.async_utils import synchronize_api
+from ._utils.grpc_utils import RETRYABLE_GRPC_STATUS_CODES
+from .client import _Client
+from .exception import FilesystemExecutionError
+
+LARGE_FILE_SIZE_LIMIT = 16 * 1024 * 1024  # 16 MiB
+READ_FILE_SIZE_LIMIT = 100 * 1024 * 1024  # 100 MiB
+
+ERROR_MAPPING = {
+    api_pb2.SystemErrorCode.SYSTEM_ERROR_CODE_UNSPECIFIED: FilesystemExecutionError,
+    api_pb2.SystemErrorCode.SYSTEM_ERROR_CODE_PERM: PermissionError,
+    api_pb2.SystemErrorCode.SYSTEM_ERROR_CODE_NOENT: FileNotFoundError,
+    api_pb2.SystemErrorCode.SYSTEM_ERROR_CODE_IO: IOError,
+    api_pb2.SystemErrorCode.SYSTEM_ERROR_CODE_NXIO: IOError,
+    api_pb2.SystemErrorCode.SYSTEM_ERROR_CODE_NOMEM: MemoryError,
+    api_pb2.SystemErrorCode.SYSTEM_ERROR_CODE_ACCES: PermissionError,
+    api_pb2.SystemErrorCode.SYSTEM_ERROR_CODE_EXIST: FileExistsError,
+    api_pb2.SystemErrorCode.SYSTEM_ERROR_CODE_NOTDIR: NotADirectoryError,
+    api_pb2.SystemErrorCode.SYSTEM_ERROR_CODE_ISDIR: IsADirectoryError,
+    api_pb2.SystemErrorCode.SYSTEM_ERROR_CODE_INVAL: OSError,
+    api_pb2.SystemErrorCode.SYSTEM_ERROR_CODE_MFILE: OSError,
+    api_pb2.SystemErrorCode.SYSTEM_ERROR_CODE_FBIG: OSError,
+    api_pb2.SystemErrorCode.SYSTEM_ERROR_CODE_NOSPC: OSError,
+}
+
+T = TypeVar("T", str, bytes)
+
+
+# The Sandbox file handling API is designed to mimic Python's io.FileIO
+# See https://github.com/python/cpython/blob/main/Lib/_pyio.py#L1459
+# Unlike io.FileIO, it also implements some higher level APIs, like `delete_bytes` and `overwrite_bytes`.
```

Comment:
Could we document the reasoning for having those specific extensions/higher level API methods. (I assume it's for efficiency reasons, not having to shuffle all of the file data back and forth to rewrite parts of a file).

Also, to make the distinction of custom vs standard clear, maybe we could move these non-standard methods to their own top level helper functions in the module? I can see the case for having them be in the FileIO class too (IDE discoverability/tab completion), but at the same time API discoverability through the web docs might be easier if these special methods are top level and not buried among the standard functions that a user wouldn't explore (since it looks exactly like io.FileIO which they already know otherwise)

Reply:
Thanks for bringing up the web docs! Moving the implementation in a helper makes sense to me; unfortunately the docs end up looking something like this: 

<img width="1232" alt="Screenshot 2024-12-11 at 5 14 55 AM" src="https://github.com/user-attachments/assets/10216e62-f885-4cd0-8db2-ff4339e47751">

where the higher-level functions are separated but pushed to the bottom (in addition, the docstrings are also duplicated between methods). Is this what you had in mind?

Reply:
That looks pretty normal to me - what would be your expected output?

What do you mean by the docstrings being duplicated between methods?

Reply:
Ah, just that the docstring on the `delete_bytes` in `modal.file_io.FileIO` and `modal.file_io.delete_bytes` is the same